airflow/io/store/path.py

```diff
@@ -0,0 +1,666 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import contextlib
+import os
+import typing
+from io import UnsupportedOperation
+from stat import S_ISLNK
+
+from fsspec.utils import stringify_path
+
+from airflow.io.store import ObjectStore, attach
+
+
+def _rewrite_info(info: dict, store: ObjectStore) -> dict:
+    info["name"] = ObjectStoragePath(info["name"], store=store)
+    return info
+
+
+class ObjectStoragePath(os.PathLike):
+    """A path-like object for object storage."""
+
+    sep: typing.ClassVar[str] = "/"
+    root_marker: typing.ClassVar[str] = "/"
+
+    __slots__ = (
+        "_store",
+        "_bucket",
+        "_key",
+        "_conn_id",
+        "_protocol",
+        "_hash",
+    )
+
+    def __init__(self, path, conn_id: str | None = None, store: ObjectStore | None = None):
+        self._conn_id = conn_id
+        self._store = store
+
+        self._hash = None
+
+        self._protocol, self._bucket, self._key = self.split_path(path)
+
+        if store:
+            self._conn_id = store.conn_id
+            self._protocol = self._protocol if self._protocol else store.protocol
+        elif self._protocol:
+            # if not self._bucket and self._protocol != "file":
+            #   raise ValueError("Cannot create path with protocol but no bucket")
+
+            self._store = attach(self._protocol, conn_id)
+
+    @classmethod
+    def split_path(cls, path) -> tuple[str, str, str]:
+        protocol = ""
+        key = ""
+
+        path = stringify_path(path)
+
+        i = path.find("://")
+        if i > 0:
+            protocol = path[:i]
+            path = path[i + 3 :]
+
+        if cls.sep not in path:
+            bucket = path
+        else:
+            bucket, key = path.split(cls.sep, 1)
+
+        # we don't care about versions etc
+        return protocol, bucket, key
+
+    def __fspath__(self):
+        return self.__str__()
+
+    def __repr__(self):
+        return f"<{type(self).__name__}('{self}')>"
+
+    def __str__(self):
+        path = (
+            f"{self._protocol}://{self._bucket}/{self._key}"
+            if self._protocol
+            else f"{self._bucket}/{self._key}"
+        )
+
+        return path
+
+    def __lt__(self, other):
+        if not isinstance(other, type(self)):
```

Comment:
It is generally more preferred to use the real class name here instead of `type(self)`, otherwise it’d be awkward to subclass. But subclass is probably not particularly useful for this class?

Reply:
Happy to use the real class name.